File: auxiliary/Thrift/src/greta/auxiliary/thrift/services/ExternalClient.java
/*
 * This file is part of the auxiliaries of Greta.
 *
 * Greta is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Greta is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Greta.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
package greta.auxiliary.thrift.services;

import greta.auxiliary.thrift.gen_java.ExternalCom;
import greta.auxiliary.thrift.gen_java.Message;
import greta.core.util.log.Logs;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;

/**
 *
 * @author Ken Prepin
 */
public abstract class ExternalClient extends Connector<ExternalClient>{

    TSocket transport;
    ExternalCom.Client client;
    ExternalClientConnectionCheckThread extClientThread;

   Message message;
   boolean newMessage;

   public ExternalClient(){
        this(DEFAULT_THRIFT_HOST,DEFAULT_THRIFT_PORT);
    }
    public ExternalClient(String host, int port){
        super(host, port);
        message = new Message();
        message.setId(ExternalComHandler.messageInitId);
        newMessage =  false;
    }
    public void updateMessage() {
        if(this.isConnected()){
            try {
                synchronized(client){
                   Message m = new Message(client.update(message.getId()));
                   if(!"empty".equals(m.getType())){
                        message = client.update(message.getId());
                   }
                }

            } catch (Exception ex) {
                System.err.println("Can not send Message. " + ex.getMessage());
                System.err.println("Check that the right Receiver is connected on " + this.getHost() + " " + this.getPort());
                System.err.println("Check that there is no remaining java.exe running");
                if (isConnected()) {
                    startConnection();
                }
            }
        } else {
            Logs.info("Thrift Sender not connected");
        }
    }

    public boolean isNewMessage(){
        return newMessage;
    }

    public Message getMessage(){
        Logs.debug("new message " + message.id +" received on the client");
        return message;
    }

    @Override
    protected void finalize() throws Throwable {
        synchronized (this) {
            if (transport != null) {
                System.out.println("Closing the simple client...");
                transport.close();
            }
        }
        super.finalize();
    }

    @Override
    public void startConnector(ExternalClient extClient) {
        try {
            synchronized (extClient) {
                extClient.transport = new TSocket(extClient.getHost(), extClient.getPort());
                extClient.transport.open();
                extClient.transport.setTimeout(1000);
                TProtocol protocol = new TBinaryProtocol(extClient.transport);
                extClient.client = new ExternalCom.Client(protocol);
                System.out.println("Connected to " + extClient.getHost() + " - " + extClient.getPort());
                System.out.println("sender to String: " + extClient.client.toString());
                if (extClient.client != null) {
                    setConnected(true);
                }
                extClientThread = new ExternalClient.ExternalClientConnectionCheckThread(this);
                extClientThread.start();
            }
        } catch (Throwable ex) {
            System.err.println("Exception " + ex.getMessage());
        }
    }
    @Override
    public void stopConnector(ExternalClient externalClient){
        if(externalClient.transport.isOpen()){
            externalClient.transport.close();
        }
        externalClient.setConnected(false);
    }
    public void send(Message m) {
        if(this.isConnected()){
            try {
                synchronized(client){
                    client.send(m);
                }

            } catch (Exception ex) {
                System.err.println("Can not send Message. " + ex.getMessage());
                System.err.println("Check that the right Receiver is connected on " + this.getHost() + " " + this.getPort());
                System.err.println("Check that there is no remaining java.exe running");
                if (isConnected()) {
                    startConnection();
                }
            }
        } else {
            Logs.info("Thrift Sender not connected");
        }
    }


    @Override
    public void setPort(int port) {
        if (getPort() != port) {
            super.setPort(port);
            startConnection();
        }
    }

    @Override
    public void setPort(String port) {
        setPort(Integer.parseInt(port));
    }

    @Override
    public void setHost(String host) {
        if (!getHost().equals(host)) {
            super.setHost(host);
            startConnection();
        }
    }
    private class ExternalClientConnectionCheckThread extends Thread {
        final ExternalClient externalClient;

        public ExternalClientConnectionCheckThread(ExternalClient externalClientToConnect){
            Logs.debug("ExternalClientConnectionCheckThread created");
            this.setDaemon(true);
            this.externalClient=externalClientToConnect;

            this.setName("ThriftExternalClientConnectionCheckThread "+externalClient.getHost()+ " "+ externalClient.getPortString());
        }

        @Override
        public void run() {
            Logs.debug("ExternalClientConnectionCheck started");
            try {
                while (externalClient.isConnected()) {
                    boolean clientStarted = false;
                    synchronized(externalClient.client){
                        clientStarted = externalClient.client.isStarted();
                    }
                    if (clientStarted) {
                        try {
                            sleep(SLEEP_TIME);
                        } catch (InterruptedException ex) {
                            Logger.getLogger(Sender.class.getName()).log(Level.SEVERE, null, ex);
                        }
                    }
                }
            } catch (Exception ex) {

                Logs.debug("ExternalClientConnectionCheckThread: sender.client.isStarted : exception = " + ex.getMessage());
                externalClient.startConnection();

            }
        }

    }
    public abstract void perform(Message m);
}

File: gretaUnity-master-1/gretaUnity-master/Assets/Scripts/ThriftImpl/services/ExternalClient.cs
using System;
using UnityEngine;
using System.Runtime.CompilerServices;
using System.Threading;


using thrift.gen_csharp;

using Thrift;
using Thrift.Protocol;
using Thrift.Transport;
using Thrift.Collections;

namespace thrift.services
{
	public abstract class ExternalClient : Connector
	{
		TSocket transport;
		ExternalCom.Client client;
		Thread extClientCheckConnectionThread;
		Thread extClientUpdateMessageThread;
		Message message;
		bool newMessage;
		int cptThreadStarted;
		bool starting;
		
		public ExternalClient (): this(DEFAULT_THRIFT_HOST,DEFAULT_THRIFT_PORT)
		{
		}

		public ExternalClient (String host, int port): base(host, port)
		{
			message = new Message ();
			message.Id = "initMessage";
			newMessage = false;	
			cptThreadStarted = 0;
		}

		~ExternalClient ()
		{
			if (transport != null) {
				Debug.Log ("Closing the simple client...");
				transport.Close ();
			}
		}

		public void updateMessage ()
		{
			//Debug.Log ("updateMessage in thread: "+Thread.CurrentThread.GetHashCode());
			if (this.isConnected ()) {
				try {
					lock (client) {
						long requestLength = DateTime.Now.Ticks/10000;
						Message m = client.update (message.Id);
						requestLength = DateTime.Now.Ticks/10000 - requestLength;
/* Debug request length !!!*/	Debug.Log (" requestLength " + requestLength);
						if (m.Type != "empty") {
							newMessage = true;
							message = m;
							//	Debug.Log("new message received on port: "+this.getPort ()+" "+m.Id +" "+m.FirstFrameNumber +" "+ m.LastFrameNumber);
						} else if (m.Type == "empty"){
							newMessage = false;
						} else {
							client.isStarted ();
							
						}
					}
				} catch (Exception ex) {
					Debug.Log ("Cannot get update from ServerToExternal on " + this.getHost () + " " + this.getPort () + " "+ex.Message + ex.StackTrace + ex.InnerException);
				//	Debug.Log ("Check that there is no remaining java.exe running");
					/*if (isConnected ()) {
						if(!this.client.isStarted ()){
							startConnection ();
						}
					}*/
					//updateMessage ();
				}
			} else {
				Debug.Log ("Thrift ExternalClient not connected");
			}
		}
    
		public bool isNewMessage ()
		{
			return newMessage;
		}
    
		public Message getMessage ()
		{
			//    Debug.Log("new message " + message.Id +" received on the client");
			return message;
		}
		public override void startConnector ()
		{
			try {
				lock (this) {
					this.transport = new TSocket (this.getHost (), this.getPort ());
					this.transport.Open ();
					this.transport.Timeout=1000;
					TProtocol protocol = new TBinaryProtocol (this.transport);
					this.client = new ExternalCom.Client (protocol);
					if (this.client != null) {
						setConnected (true);
					//	Debug.Log ("Connected to " + this.getHost () + " - " + this.getPort ());
						/*     extClientCheckConnectionThread = new Thread(new ThreadStart(ExternalClientConnectionCheckThread));
					extClientCheckConnectionThread.IsBackground = true; 
	                extClientCheckConnectionThread.Start();*/
						
						if(extClientUpdateMessageThread==null){
							extClientUpdateMessageThread = new Thread (new ThreadStart (ExternalClientUpdateMessageThread));
							extClientUpdateMessageThread.IsBackground = true; 
							extClientUpdateMessageThread.Start ();
						}
						
					}
				}
			} catch (Exception ex) {
				Console.Error.WriteLine ("Exception " + ex.Message);
			}
		}
    
		public override void stopConnector ()
		{
			if(this.isOnConnection()  || this.isConnected ()){
		//	Debug.Log ("stopConnector this.transport.IsOpen " + this.transport.IsOpen);
				if (this.transport.IsOpen) {
					this.transport.Close ();
				}
					
				this.setConnected(false);
			}
		//	Debug.Log ("stopConnector this.transport.IsOpen " + this.transport.IsOpen);
		}

		public void send (Message m)
		{
			try {
				lock (client) {
					client.send (m);
				}
			} catch (Exception ex) {
				Debug.Log ("Cannot send Message on " + this.getHost () + " " + this.getPort ()+ " " + ex.StackTrace);
			//	Debug.Log ("Check that there is no remaining java.exe running");
				if (isConnected ()) {
					startConnection ();
				}
			}
		}

		public new void setPort (int port)
		{
			if (getPort () != port) {
				base.setPort (port);
			}
		}

		public new void setPort (String port)
		{
			setPort (Convert.ToInt16 (port));
		}

		public new void setHost (String host)
		{
			if (!getHost ().Equals (host)) {
				base.setHost (host);
			}
  
		}

		public void ExternalClientUpdateMessageThread ()
		{
        
			cptThreadStarted ++;
		//	Debug.Log ("externalClientUpdateMessageThread started number: " + cptThreadStarted);
			while (Thread.CurrentThread==extClientUpdateMessageThread) {
				if(isConnected()){
						updateMessage ();
						if (isNewMessage ()) {
							perform (this.getMessage ());
						}
				}
				try {
					Thread.Sleep (100);
				} catch (Exception ex) {
					Debug.Log (ex);
				}
                

			}
			cptThreadStarted --;
		//	Debug.Log ("externalClientUpdateMessageThread stopping number: " + cptThreadStarted);        
				
		}

		public abstract void perform (Message m);
		
		public void ExternalClientConnectionCheckThread ()
		{
        
		//	Debug.Log ("ExternalClientConnectionCheckThread start");
			try {
				while (isConnected()) {
					bool clientStarted = false;
					lock (this.client) {
						clientStarted = this.client.isStarted ();
					}
					if (clientStarted) {
						try {
							Thread.Sleep (100);
						} catch (Exception ex) {
							Debug.Log (ex);
						}
					}
				}
			} catch (Exception e) {
				Debug.Log (e);
				this.startConnection ();

			}
		}
	

	}
	

}